File: solutions/go/queen-attack/1/queen_attack.go
package queenattack

import "errors"

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func CanQueenAttack(whitePosition, blackPosition string) (bool, error) {
	if whitePosition == blackPosition {
		return false, errors.New("same square")
	}

	if whitePosition == "" || blackPosition == "" {
		return false, errors.New("empty position")
	}

	whiteCoords := make([]int, 2)
	blackCoords := make([]int, 2)

	whiteCoords[0] = int(whitePosition[0] - 'a')
	whiteCoords[1] = int(whitePosition[1] - '1')

	blackCoords[0] = int(blackPosition[0] - 'a')
	blackCoords[1] = int(blackPosition[1] - '1')

	if whiteCoords[0] < 0 || whiteCoords[0] > 7 || whiteCoords[1] < 0 || whiteCoords[1] > 7 {
		return false, errors.New("invalid position")
	}
	if blackCoords[0] < 0 || blackCoords[0] > 7 || blackCoords[1] < 0 || blackCoords[1] > 7 {
		return false, errors.New("invalid position")
	}

	dx := abs(whiteCoords[0] - blackCoords[0])
	dy := abs(whiteCoords[1] - blackCoords[1])

	if dx == 0 && dy == 0 {
		return false, nil
	}

	if dx == dy || dx == 0 || dy == 0 {
		return true, nil
	}

	return false, nil
}

File: solutions/typescript/queen-attack/1/queen-attack.ts
type Position = readonly [number, number]

type Positions = {
  white: Position
  black: Position
}
export class QueenAttack {
  public readonly black: Position
  public readonly white: Position

  // white: [whiteRow, whiteColumn]
  // black: [blackRow, blackColumn]
  constructor({white = [7, 3], black = [0, 3]}: Partial<Positions> = {}) {
    const onBoard = ([y, x]: Position) => x >= 0 && x <= 7 && y >= 0 && y <= 7;
    if (!onBoard(white) || !onBoard(black)) {
      throw new Error('Queen must be placed on the board');
    }
    if (white[0] === black[0] && white[1] === black[1]) {
      throw new Error('Queens cannot share the same space');
    }

    this.white = white;
    this.black = black;
  }

  toString() {
    let board: string[][] = Array(8).fill(null).map(_ => Array(8).fill('_'));
    const [whiteRow, whiteColumn] = this.white;
    const [blackRow, blackColumn] = this.black;
    board[whiteRow][whiteColumn] = 'W';
    board[blackRow][blackColumn] = 'B';

    return board.map(row => row.join(' ')).join('\n');
  }

  get canAttack() {
    const [whiteRow, whiteColumn] = this.white;
    const [blackRow, blackColumn] = this.black;

    return (
      whiteRow === blackRow ||
      whiteColumn === blackColumn ||
      Math.abs(whiteRow - blackRow) === Math.abs(whiteColumn - blackColumn) // diagonal
    );
  }
}